src/layers.rs

```diff
@@ -42,42 +43,145 @@ impl LayerTile {
     }
 }
 
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Clone, PartialEq, Debug)]
+pub enum LayerType {
+    TileLayer(TileLayer),
+    ObjectLayer(ObjectLayer),
+    ImageLayer(ImageLayer),
+    // TODO: Support group layers
+}
+
+impl LayerType {
+    pub fn as_tile_layer(&self) -> Option<&TileLayer> {
+        match self {
+            LayerType::TileLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_tile_layer_mut(&mut self) -> Option<&mut TileLayer> {
+        match self {
+            LayerType::TileLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+
+    pub fn as_object_layer(&self) -> Option<&ObjectLayer> {
+        match self {
+            LayerType::ObjectLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_object_layer_mut(&mut self) -> Option<&mut ObjectLayer> {
+        match self {
+            LayerType::ObjectLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+
+    pub fn as_image_layer(&self) -> Option<&ImageLayer> {
+        match self {
+            LayerType::ImageLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_image_layer_mut(&mut self) -> Option<&mut ImageLayer> {
+        match self {
+            LayerType::ImageLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub(crate) enum LayerTag {
+    TileLayer,
+    ObjectLayer,
+    ImageLayer,
+}
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct Layer {
     pub name: String,
-    pub width: u32,
-    pub height: u32,
-    pub opacity: f32,
+    pub id: u32,
     pub visible: bool,
     pub offset_x: f32,
     pub offset_y: f32,
-    /// The tiles are arranged in rows. Each tile is a number which can be used
-    ///  to find which tileset it belongs to and can then be rendered.
-    pub tiles: LayerData,
+    pub opacity: f32,
     pub properties: Properties,
-    pub layer_index: u32,
-    /// The ID of the layer, as shown in the editor.
-    /// Layer ID stays the same even if layers are reordered or modified in the editor.
-    pub id: u32,
+    pub ty: LayerType,
 }
 
 impl Layer {
     pub(crate) fn new<R: Read>(
         parser: &mut EventReader<R>,
         attrs: Vec<OwnedAttribute>,
-        layer_index: u32,
+        tag: LayerTag,
         infinite: bool,
-    ) -> Result<Layer, TiledError> {
-        let ((o, v, ox, oy, n, id), (w, h)) = get_attrs!(
+        path_relative_to: Option<&Path>,
+    ) -> Result<Self, TiledError> {
+        let ((opacity, visible, offset_x, offset_y, name), id) = get_attrs!(
             attrs,
             optionals: [
                 ("opacity", opacity, |v:String| v.parse().ok()),
                 ("visible", visible, |v:String| v.parse().ok().map(|x:i32| x == 1)),
                 ("offsetx", offset_x, |v:String| v.parse().ok()),
                 ("offsety", offset_y, |v:String| v.parse().ok()),
                 ("name", name, |v| Some(v)),
+            ],
+            required: [
```

Comment:
We just made the `id` not required in aa8214abe2e33dc9988e1255eb7dceab604d9447. Even if Tiled will always write it, I don't see a reason to require it. Tiled itself for example needs to be able to load maps without these, for backwards compatibility.